crates/typst-macros/src/elem.rs

```diff
@@ -480,88 +472,904 @@ fn create_param_info(field: &Field) -> TokenStream {
+                    _ => Err(()),
+                }
+            }
+        }
+
+        impl ::std::fmt::Display for #enum_ident {
+            fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
+                match self {
+                    #(
+                        Self::#field_variants => f.write_str(#field_names),
+                    )*
+                    Self::Label => f.write_str("label"),
+                }
+            }
+        }
+
+        impl ::std::str::FromStr for #enum_ident {
+            type Err = ();
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                match s {
+                    #(
+                        #field_names => Ok(Self::#field_variants),
+                    )*
+                    "label" => Ok(Self::Label),
+                    _ => Err(()),
+                }
+            }
+        }
+    }
+}
+
+/// Create the `new` function for the element.
+fn create_new_func(element: &Elem) -> TokenStream {
+    let relevant = element
+        .fields
+        .iter()
+        .filter(|field| !field.external && !field.synthesized && field.inherent());
+    let params = relevant.clone().map(|Field { ident, ty, .. }| {
+        quote! { #ident: #ty }
+    });
+    let required = relevant.map(|Field { ident, .. }| {
+        quote! { #ident }
+    });
+    let defaults = element
+        .fields
+        .iter()
+        .filter(|field| !field.external && (field.synthesized || !field.inherent()))
+        .map(|Field { ident, .. }| {
+            quote! { #ident: None }
+        });
+
+    let label_and_location = element.without_capability("Unlabellable", || {
+        quote! {
+            location: None,
+            label: None,
+            prepared: false,
+        }
+    });
+
+    quote! {
+        /// Create a new element.
+        pub fn new(#(#params),*) -> Self {
+            Self {
+                span: ::typst::syntax::Span::detached(),
+                #label_and_location
+                guards: ::std::vec::Vec::with_capacity(0),
+                #(#required,)*
+                #(#defaults,)*
+            }
+        }
+    }
+}
+
+/// The `..` in `#[elem(..)]`.
+struct Meta {
+    scope: bool,
+    name: Option<String>,
+    title: Option<String>,
+    keywords: Vec<String>,
+    capabilities: Vec<Ident>,
+}
+
+impl Parse for Meta {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            scope: parse_flag::<kw::scope>(input)?,
+            name: parse_string::<kw::name>(input)?,
+            title: parse_string::<kw::title>(input)?,
+            keywords: parse_string_array::<kw::keywords>(input)?,
+            capabilities: Punctuated::<Ident, Token![,]>::parse_terminated(input)?
+                .into_iter()
+                .collect(),
+        })
+    }
+}
+
+/// Parse details about the element from its struct definition.
+fn parse(stream: TokenStream, body: &syn::ItemStruct) -> Result<Elem> {
```

Comment:
I'd prefer if the `Meta` struct and `parse` function came earlier in the file as before, since the logical order is parse -> create.

Reply:
That's how it was before iirc that's why I kept it as is, will change